File: examples/3PMD/prediction/NN/classification/inference.py
import torch
import os
from model import NeuralNet
import argparse
import numpy as np

PROJECT_PATH = os.path.dirname(os.path.abspath(__file__))
SAVE_PATH = os.path.join(PROJECT_PATH, "models")
SAVE_NAME = os.path.join(SAVE_PATH, "model.pt")

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device = "cpu"
print("Using device:", device)

def predict_top_n(model, X_input, N=3):
    model.eval()
    X_input = X_input.to(device) 

    with torch.no_grad():
        # Directly use the model's output (already probabilities if softmax is in the model)
        probs = model(X_input)
        probs = torch.softmax(probs, dim=1)
        top_n_probs, top_n_classes = torch.topk(probs, N, dim=1)

    return top_n_classes.cpu().numpy(), top_n_probs.cpu().numpy()

def preprocess_input(data):
    """
    Preprocess input data by computing cosine and sine for angle features.
    """
    angles = ['theta_i', 'theta_f', 'alpha_m', 'alpha_f']
    processed_data = []
    for i, value in enumerate(data):
        if angles[i % len(angles)] in angles:
            angle_rad = np.radians(value)
            processed_data.append(np.cos(angle_rad))
            processed_data.append(np.sin(angle_rad))
        else:
            processed_data.append(value)
    return processed_data

def preprocess_data(data):
    """
    Preprocess data by computing cosine and sine for angle features.
    """
    angles = ['theta_i', 'theta_f', 'alpha_m', 'alpha_f']
    for angle in angles:
        data[f'{angle}_cos'] = np.cos(np.radians(data[angle]))
        data[f'{angle}_sin'] = np.sin(np.radians(data[angle]))
    return data

if __name__ == "__main__":
    
    parser = argparse.ArgumentParser(description='Inference script for the Neural Network model')
    parser.add_argument('--input', type=str, help='Input data for prediction', required=True)
    args = parser.parse_args()
    
    model = NeuralNet()  
    model.load_state_dict(torch.load(SAVE_NAME, map_location=device, weights_only=True))
    model.to(device)

    if args.input.endswith(".csv"):
        import pandas as pd
        data = pd.read_csv(args.input, sep='\s+')
        
        original_features = ['kmax', 'theta_i', 'theta_f', 'alpha_m', 'alpha_f', 'id_man_comb']
        data = data[original_features]

        data = preprocess_data(data)

        # Update feature list to include cosine and sine of angles
        features = ['kmax', 'theta_i_cos', 'theta_i_sin', 'theta_f_cos', 'theta_f_sin',
                    'alpha_m_cos', 'alpha_m_sin', 'alpha_f_cos', 'alpha_f_sin']
        target = 'id_man_comb'

        # Preprocess data
        X = data[features].values
        y = data[target].values

        tot = 0 
        correct = 0

        for entry, label in zip(X, y):
            data = torch.tensor([entry], dtype=torch.float32)
            top_classes, top_probs = predict_top_n(model, data, N=3)
            # print("Input:", entry)
            # print("Target Class:", y[tot])
            # print("Top-N Predicted Classes:", top_classes)
            print("Top-N Probabilities:", top_probs)
            # print()

            if y[tot] in top_classes[0]:
                correct += 1
            tot += 1

        print("Accuracy:", correct / tot * 100.0)

    else:
        # Transform string to list of floats
        args.input = args.input.split(',')
        args.input = [float(i) for i in args.input]

        # Preprocess input to compute cosine and sine for angles
        args.input = preprocess_input(args.input)
        
        data = torch.tensor([args.input], dtype=torch.float32)
        
    top_classes, top_probs = predict_top_n(model, data, N=5)

    print("Top-N Predicted Classes:", top_classes)
    print("Top-N Probabilities:", top_probs)

File: examples/3PMD/prediction/NN/classification/inference.cpp
#include <torch/script.h>
#include <torch/torch.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <string>
#include <Eigen/Dense>
#include <nlohmann/json.hpp> // Include a JSON library like nlohmann/json
#include <fstream>

torch::Device device(torch::cuda::is_available() ? torch::kCUDA : torch::kCPU);

std::vector<float> parseInputString(const std::string& input) {
    std::vector<float> values;
    std::stringstream ss(input);
    std::string item;
    while (std::getline(ss, item, ',')) {
        values.push_back(std::stof(item));
    }
    return values;
}

Eigen::VectorXf applyScaler(const Eigen::VectorXf& input, const Eigen::VectorXf& mean, const Eigen::VectorXf& scale) {
    return (input - mean).cwiseQuotient(scale);
}

std::pair<Eigen::VectorXf, Eigen::VectorXf> loadScaler(const std::string& scaler_json_path) {
    std::ifstream file(scaler_json_path);
    if (!file) {
        throw std::runtime_error("Error opening scaler JSON file: " + scaler_json_path);
    }

    nlohmann::json scaler_json;
    file >> scaler_json;

    std::vector<float> mean_vec = scaler_json["mean"];
    std::vector<float> scale_vec = scaler_json["scale"];

    Eigen::VectorXf mean = Eigen::Map<Eigen::VectorXf>(mean_vec.data(), mean_vec.size());
    Eigen::VectorXf scale = Eigen::Map<Eigen::VectorXf>(scale_vec.data(), scale_vec.size());

    return {mean, scale};
}

std::tuple<torch::Tensor, torch::Tensor> predict(torch::jit::script::Module& model, torch::Tensor input, int N) {
    model.to(device);
    model.eval();
    input = input.to(device);
    
    torch::Tensor output = model.forward({input}).toTensor();
    torch::Tensor probabilities = torch::softmax(output, 1);
    torch::Tensor top_probs, top_indices;
    std::tie(top_probs, top_indices) = torch::topk(probabilities, N, 1);

    return {top_indices.cpu(), top_probs.cpu()};
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <input_values_or_csv_file>" << std::endl;
        return 1;
    }

    std::string input_arg = argv[1];
    std::string scaler_json_path = "models/model1_scaler.json"; // Path to the scaler JSON file

    Eigen::VectorXf mean, scale;
    try {
        std::tie(mean, scale) = loadScaler(scaler_json_path);
    } catch (const std::exception& e) {
        std::cerr << "Error loading scaler: " << e.what() << std::endl;
        return -1;
    }

    torch::jit::script::Module model;

    try {
        model = torch::jit::load("models/model_cpp.pt", device);
    } catch (const c10::Error& e) {
        std::cerr << "Error loading the model." << std::endl;
        std::cerr << e.what() << std::endl;
        return -1;
    }

    if (input_arg.find(".csv") != std::string::npos) {
        std::ifstream file(input_arg);
        if (!file) {
            std::cerr << "Error opening file: " << input_arg << std::endl;
            return -1;
        }
        
        float kmax, theta_i, theta_f, alpha_m, alpha_f;
        int id_man_comb;

        std::vector<std::vector<float>> data;
        std::vector<int> labels;

        std::string tmp;
        std::getline(file, tmp); // Skip header

        int counter = 0;

        while(file >> kmax >> theta_i >> theta_f >> alpha_m >> alpha_f >> id_man_comb) {
            Eigen::VectorXf input(5);
            input << kmax, theta_i, theta_f, alpha_m, alpha_f;
            Eigen::VectorXf scaled_input = applyScaler(input, mean, scale);

            data.push_back(std::vector<float>(scaled_input.data(), scaled_input.data() + scaled_input.size()));
            labels.push_back(id_man_comb);
            // if (counter == 5) break;
            counter++;
        }

        torch::Tensor input_tensor = torch::from_blob(data.data(), {(long)data.size(), (long)data[0].size()}, torch::kFloat32);
        
        auto [top_classes, top_probs] = predict(model, input_tensor, 1);
        // std::cout << "Top-N Predicted Classes: " << std::endl << top_classes << std::endl;
        // std::cout << "Top-N Probabilities: " << std::endl << top_probs << std::endl;

        int correct = 0;
        int total = 0;
        for (int i = 0; i < labels.size(); i++) {
            for (int j = 0; j < top_classes.size(1); j++) {
                if (labels[i] == top_classes[i][j].item<int>()) {
                    correct++;
                    break;
                }
            }
        }
        std::cout << "Accuracy: " << (float)correct / labels.size() << std::endl;

    } else {
        std::vector<float> input_values = parseInputString(input_arg);
        Eigen::VectorXf input = Eigen::Map<Eigen::VectorXf>(input_values.data(), input_values.size());
        Eigen::VectorXf scaled_input = applyScaler(input, mean, scale);

        torch::Tensor input_tensor = torch::from_blob(scaled_input.data(), {(long)scaled_input.size()}, torch::kFloat32).unsqueeze(0);
        
        auto [top_classes, top_probs] = predict(model, input_tensor, 5);
        std::cout << "Top-N Predicted Classes: " << top_classes << std::endl;
        std::cout << "Top-N Probabilities: " << top_probs << std::endl;
    }

    return 0;
}
